### PR TITLE
[TE] Endpoints for bulk on-boarding, subscription and tagging of metrics

### PR DESCRIPTION
**Pull request by akshayrai**

### CODE REVIEW

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/tasks/FunctionCreationOnboardingTask.java [hunk 1]
```diff
@@ -298,6 +322,21 @@ public void run() {
     executionContext.setExecutionResult(ALERT_CONFIG, alertConfig);
   }
 
+  private AnomalyFunctionDTO getDefaultFunctionSpec(boolean isLegacy, TimeGranularity dataGranularity) {
+    AnomalyFunctionDTO anomalyFunctionSpec = new AnomalyFunctionDTO();
+    if (isLegacy) {
+      anomalyFunctionSpec = getDefaultFunctionSpecByTimeGranularity(dataGranularity);
+    } else {
+      anomalyFunctionSpec.setType("SPLINE_REGRESSION_WRAPPER");
```
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/tasks/FunctionCreationOnboardingTask.java]: does this work when time granularity is minutes or hours?

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/AnomalyResource.java [hunk 2]
```diff
@@ -106,6 +117,7 @@
   private static final String DEFAULT_FUNCTION_TYPE = "WEEK_OVER_WEEK_RULE";
 
   private AnomalyFunctionManager anomalyFunctionDAO;
+  private TaskManager taskDAO;
```
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/AnomalyResource.java]: are you using this?

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/AnomalyResource.java [hunk 3]
```diff
@@ -890,4 +906,4 @@ public void updateAnomalyMergedResultFeedback(@PathParam("anomaly_merged_result_
     }
   }
 
-}
+}
```
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/AnomalyResource.java]: nit: new line

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/EmailResource.java [hunk 4]
```diff
@@ -181,6 +187,81 @@ public Response updateAlertConfigHolidayCron(@NotNull @PathParam("id") Long id,
...
+  @Path("{id}/subscribe")
+  @ApiOperation("Endpoint for subscribing to an alert/alerts")
+  public Response subscribeAlert(
+      @PathParam("id") Long id,
+      @QueryParam("functionId") String functionIds,
+      @QueryParam("functionName") String functionName,
+      @QueryParam("metric") String metric,
+      @QueryParam("dataset") String dataset,
+      @DefaultValue("") @QueryParam("functionPrefix") String functionPrefix) throws Exception {
+    if (org.apache.commons.lang3.StringUtils.isBlank(functionIds) && org.apache.commons.lang3.StringUtils.isBlank(functionName)
```
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/EmailResource.java]: import the base namespace. Both lang and lang3 should work here anyways

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/EmailResource.java [hunk 5]
```diff
@@ -181,6 +187,81 @@ public Response updateAlertConfigHolidayCron(@NotNull @PathParam("id") Long id,
...
+    if (org.apache.commons.lang3.StringUtils.isNotBlank(functionIds)) {
+      for (String functionId : functionIds.split(",")) {
+        AnomalyFunctionDTO anomalyFunction = anomalyDAO.findById(Long.valueOf(functionId));
+        if (anomalyFunction != null) {
+          functionIdList.add(Long.valueOf(functionId));
+        } else {
+          responseMessage.put("skipped function " + functionId, "Cannot be found!");
+        }
+      }
+    } else if (org.apache.commons.lang3.StringUtils.isNotBlank(functionName)) {
```
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/EmailResource.java]: same

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/EmailResource.java [hunk 6]
```diff
@@ -181,6 +187,81 @@ public Response updateAlertConfigHolidayCron(@NotNull @PathParam("id") Long id,
...
+    Map<String, String> responseMessage = new HashMap<>();
+
+    AlertConfigDTO alertConfigDTO = alertDAO.findById(id);
+    if (alertConfigDTO == null) {
+      responseMessage.put("message", "cannot find the alert group configuration entry " + id + ".");
+      return Response.status(Response.Status.BAD_REQUEST).entity(responseMessage).build();
+    }
+
+    final List<Long> functionIdList = new ArrayList<>();
+    if (org.apache.commons.lang3.StringUtils.isNotBlank(functionIds)) {
```
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/EmailResource.java]: same

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/MetricConfigResource.java [hunk 7]
```diff
@@ -219,6 +225,36 @@ public Long createMetricConfig(String payload) {
     return id;
   }
 
+  @PUT
+  @Path("{id}/create-tag")
+  @ApiOperation("Endpoint for tagging a metric")
+  public Response createMetricTag(
+      @PathParam("id") Long id,
+      @NotNull @QueryParam("tag") String tag) throws Exception {
+    if (org.apache.commons.lang3.StringUtils.isBlank(tag)) {
```
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/MetricConfigResource.java]: same

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/datalayer/bao/AnomalyFunctionManager.java [hunk 8]
```diff
@@ -25,6 +25,8 @@
 
   List<AnomalyFunctionDTO> findAllByCollection(String collection);
 
+  List<AnomalyFunctionDTO> findAllByMetricAndCollection(String metric, String collection);
```
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/datalayer/bao/AnomalyFunctionManager.java]: meta: please consider using *findByPredicate()* in the code directly. The DAO interfaces are already overloaded with calls that are used only once.

Comment (akshayrai) [reply thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/datalayer/bao/AnomalyFunctionManager.java]: Good point. Switched to findByPredicate

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/EmailResource.java (hunk 4 above)
Comment (akshayrai) [reply thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/EmailResource.java]: Aah! that must have happened while copy-pasting it to a different resource file. Will fix it.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/AnomalyResource.java (hunk 2 above)
Comment (akshayrai) [reply thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/AnomalyResource.java]: Removed

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/tasks/FunctionCreationOnboardingTask.java (hunk 1 above)
Comment (akshayrai) [reply thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/tasks/FunctionCreationOnboardingTask.java]: This endpoint works only for daily metrics until we have the new algorithms supporting MIN and HOUR granularity.

Comment (apucher) [reply thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/tasks/FunctionCreationOnboardingTask.java]: ok I'll throw an Exception on non-daily granularity then.

**Pull request merged**